### PR TITLE
eth/catalyst: implement engine_getPayloadBodies* methods

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
Spec: ~~https://github.com/ethereum/execution-apis/pull/146/files~~ https://github.com/ethereum/execution-apis/pull/218/files

Will undraft + add tests once the spec is merged

### CODE REVIEW

File: eth/catalyst/api.go [hunk 1]
```diff
@@ -682,3 +682,21 @@ func (api *ConsensusAPI) heartbeat() {
...
 	}
 }
+
+// GetPayloadBodiesV1 implements engine_getPayloadBodiesV1 which allows for retrieval of a list
+// of block bodies by the engine api.
+func (api *ConsensusAPI) GetPayloadBodiesV1(hashes []common.Hash) ([]*types.Body, error) {
+	var bodies []*types.Body
+	for _, hash := range hashes {
+		body := api.eth.BlockChain().GetBody(hash)
+		if body == nil {
```
Comment (tynes) [inline eth/catalyst/api.go:692]: This doesn't seem to follow the spec? Would need to `continue` when `body == nil` 

> Client software **MUST** skip the payload body in the response array if the data of this body is missing. For instance, if the request is `[A.block_hash, B.block_hash, C.block_hash]` and client software has data of payloads `A` and `C`, but doesn't have data of `B`, the response **MUST** be `[A.body, C.body]`.

**Comment by MariusVanDerWijden** [comment]
Seems like the general consensus is to use https://github.com/ethereum/execution-apis/pull/218/files instead so I updated the implementation

**Comment by MariusVanDerWijden** [comment]
Spec changed again to https://github.com/ethereum/execution-apis/pull/352, will update the implementation to mathc

**Comment by MariusVanDerWijden** [comment]
Need to add the methods to `engine_exchangeCapabilities` once thats merged

**Comment by holiman** [comment]
Sorry, tried to fix it in the web editor, but it got screwed up :/

**Comment by MariusVanDerWijden** [comment]
It's fine, I rebased it now, should be okay

File: eth/catalyst/api.go [hunk 2]
```diff
@@ -756,3 +758,55 @@ func (api *ConsensusAPI) heartbeat() {
...
+			bodies = append(bodies, nil)
+			continue
+		}
+		// We only want to return the transactions, not uncles
+		if len(body.Uncles) != 0 {
+			body.Uncles = make([]*types.Header, 0)
+		}
+		bodies = append(bodies, body)
+	}
```
Comment (holiman) [inline eth/catalyst/api.go:782]:
```suggestion
	var bodies = make([]*types.Body, len(hashes))
	for i, hash := range hashes {
		block := api.eth.BlockChain().GetBlockByHash(hash)
		if block == nil {
			continue
		}
		body := block.Body()
		// We only want to return the transactions, not uncles
		if len(body.Uncles) != 0 {
			body.Uncles = make([]*types.Header, 0)
		}
		bodies[i] = body
	}
```

File: eth/catalyst/api.go [hunk 3]
```diff
@@ -756,3 +758,55 @@ func (api *ConsensusAPI) heartbeat() {
...
+}
+
+// GetPayloadBodiesByRangeV1 implements engine_getPayloadBodiesByRangeV1 which allows for retrieval of a range
+// of block bodies by the engine api.
+func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count uint64) []*types.Body {
+	var bodies []*types.Body
+	if api.eth.BlockChain().CurrentBlock().NumberU64() < start {
+		// Return [] if the requested range is past our latest block
+		return bodies
+	}
```
Comment (holiman) [inline eth/catalyst/api.go:793]:
```suggestion
        var bodies []*types.Body
	if api.eth.BlockChain().CurrentBlock().NumberU64() < start {
		// Return [] if the requested range is past our latest block
		return 
	}
	bodies = make([]*types.Body, count)
```
If you do this, you don't have to append `nil` later on

File: eth/catalyst/api.go [hunk 4]
```diff
@@ -756,3 +758,55 @@ func (api *ConsensusAPI) heartbeat() {
...
+		if block == nil {
+			bodies = append(bodies, nil)
+			continue
+		}
+		body := block.Body()
+		if body == nil {
+			bodies = append(bodies, nil)
+			continue
+		}
+		// We only want to return the transactions, not uncles
```
Comment (holiman) [inline eth/catalyst/api.go:805]:
```suggestion
		// We only want to return the transactions and withdrawals, not uncles
```
...  right?

File: eth/catalyst/api.go [hunk 5]
```diff
@@ -756,3 +758,55 @@ func (api *ConsensusAPI) heartbeat() {
...
+		return bodies
+	}
+	for i := uint64(0); i < count; i++ {
+		block := api.eth.BlockChain().GetBlockByNumber(start + i)
+		if block == nil {
+			bodies = append(bodies, nil)
+			continue
+		}
+		body := block.Body()
+		if body == nil {
```
Comment (holiman) [inline eth/catalyst/api.go:801]: This check is a bit too defensive -- the `Body()` can never return nil afaict:
```
// Body returns the non-header content of the block.
func (b *Block) Body() *Body { return &Body{b.transactions, b.uncles, b.withdrawals} }

```

**Review by holiman — CHANGES_REQUESTED** [review]
A few suggestions

**Comment by holiman** [comment]
The spec says, about the return: 

> This structure contains a body of an execution payload. The fields are encoded as follows:
>
>    transactions: Array of DATA - Array of transaction objects, each object is a byte list (DATA) representing TransactionType || TransactionPayload or LegacyTransaction as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)

But afaict, we return something lke 
```
{
  "Transactions": [],
  "Uncles": [],
  "Withdrawals": null
}
```
And if there are transactions, we return them in object-form, like: 
```
{
  "Transactions": [
    {
      "type": "0x0",
      "nonce": "0x0",
      "gasPrice": "0xa",
      "maxPriorityFeePerGas": null,
      "maxFeePerGas": null,
      "gas": "0xc350",
      "value": "0xa",
      "input": "0x",
      "v": "0x1b",
      "r": "0x9bea4c4daac7c7c52e093e6a4c35dbbcf8856f1af7b059ba20253e70848d094f",
      "s": "0x8a8fae537ce25ed8cb5af9adac3f141af69bd515bd2ba031522df09b97dd72b1",
      "to": "0x095e7baea6a6c7c4c2dfeb977efac326af552d87",
      "hash": "0x77b19baa4de67e45a7b26e4a220bccdbb6731885aa9927064e239ca232023215"
    },
```

So, iiuc, this should change: 

- [ ] `Transactions` -> `transactions` json
- [ ] drop Withdrawals (?) and Uncles (probably just define a new return type)
- [ ] return transactions as `a byte list (DATA) ` (?) 

Maybe I got it wrong? Double check I guess...

**Comment by holiman** [comment]
MAybe I read the wrong spec. Seems to be this instead, so both `transactions` and `withdrawals`:

This structure contains a body of an execution payload. The fields are encoded as follows:
- `transactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
- `withdrawals`: `Array of WithdrawalV1` - Array of withdrawals, each object is an `OBJECT` containing the fields of a `WithdrawalV1` structure.

**Comment by marioevz** [comment]
I agree with Martin's comment above, I tested this in the new hive tests and it should return underscore names for the json (I.e. `transactions` and `withdrawals`), and the transactions must be in hexadecimal raw form.

**Comment by holiman** [comment]
I pushed a fix to this now. It still doesn't use the most "optimal" internal API (`chain.getBodyRLP`), but I think it is correct now.

File: core/beacon/types.go [hunk 6]
```diff
@@ -229,3 +229,8 @@ func BlockToExecutableData(block *types.Block, fees *big.Int) *ExecutionPayloadE
 	}
 	return &ExecutionPayloadEnvelope{ExecutionPayload: data, BlockValue: fees}
 }
+
+type ExecutionPayloadBodyV1 struct {
```
Comment (holiman) [inline core/beacon/types.go:233]:
```suggestion
// ExecutionPayloadBodyV1 is used in the response to GetPayloadBodiesByHashV1 and GetPayloadBodiesByRangeV1
type ExecutionPayloadBodyV1 struct {
```

File: eth/catalyst/api.go [hunk 7]
```diff
@@ -756,3 +758,52 @@ func (api *ConsensusAPI) heartbeat() {
...
+		// Return [] if the requested range is past our latest block
+		return []*beacon.ExecutionPayloadBodyV1{}
+	}
+	bodies := make([]*beacon.ExecutionPayloadBodyV1, count)
+	for i := uint64(0); i < count; i++ {
+		block := api.eth.BlockChain().GetBlockByNumber(start + i)
+		if block == nil {
+			continue
+		}
+		bodies[i] = getBody(block)
```
Comment (holiman) [inline eth/catalyst/api.go:789]: You can move the `block==nil`-check inside `getBody`, and make these loops even smaller

**Review by holiman — APPROVED** [review]
LGTM, one nit and one suggestion

**Comment by marioevz** [comment]
Another small thing is that post-Shanghai blocks without withdrawals return nil withdrawals in the `ExecutionPayloadBodyV1` object response, but it should return `"withdrawals": []` instead.

**Comment by marioevz** [comment]
Another finding is that requesting a count that extends beyond the head block number results in getting a `null` filled array.
E.g.
```
head.Number==32

request:
{start==32, count==2}

current behavior:
[block32, null]

expected:
[block32]
```

On the other hand, using a start value greater than the head block number works as expected (returning empty array).

**Comment by holiman** [comment]
> 1. Client software **MUST** place `nil` in the response array for any blocks that have been pruned or where the request extends past the current latest known block.

So we should IMO always return `count` values, 


> On the other hand, using a start value greater than the head block number works as expected (returning empty array).

As I read the spec, this seems to be erroneous behaviour... ?

**Comment by holiman** [comment]
Ah wait, I was on the wrong spec. So this is the one to look at: https://github.com/ethereum/execution-apis/blob/main/src/engine/shanghai.md#specification-4 . 

> Client software MUST place null in the response array for unavailable blocks which numbers are lower than a number of the latest known block. Client software MUST NOT return trailing null values if the request extends past the current latest known block.

So for anything past `head`, just cut the reply short.

File: eth/catalyst/api.go [hunk 8]
```diff
@@ -756,3 +758,50 @@ func (api *ConsensusAPI) heartbeat() {
...
+
+// GetPayloadBodiesByRangeV1 implements engine_getPayloadBodiesByRangeV1 which allows for retrieval of a range
+// of block bodies by the engine api.
+func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count uint64) []*beacon.ExecutionPayloadBodyV1 {
+	if api.eth.BlockChain().CurrentBlock().NumberU64() < start {
+		// Return [] if the requested range is past our latest block
+		return []*beacon.ExecutionPayloadBodyV1{}
+	}
+	bodies := make([]*beacon.ExecutionPayloadBodyV1, count)
+	for i := uint64(0); i < count; i++ {
```
Comment (holiman) [inline eth/catalyst/api.go:781]:
```suggestion
        end := start + count
        last := api.eth.BlockChain().CurrentBlock().NumberU64()
        if end > last {
            end = last
        }
        var bodies []*beacon.ExecutionPayloadBodyV1{}
	for i := start; i < end i++ {
		block := api.eth.BlockChain().GetBlockByNumber(start + i)
		bodies = append(bodies, getBody(block))
```
Something like that?

Comment (MariusVanDerWijden) [reply eth/catalyst/api.go:781]: I just did exactly the same again^^

**Comment by MariusVanDerWijden** [comment]
@marioevz @holiman One thing I don't get is why we should limit start to >= 1. With that we're not able to return the genesis body

File: eth/catalyst/api.go (hunk 8 above)
Comment (holiman) [reply eth/catalyst/api.go:781]: But this time I didn't commit, so the glory is all yours!

**Comment by holiman** [comment]
>  With that we're not able to return the genesis body

Me neither. But seems kind of moot anyway, the genesis body is always empty

**Comment by MariusVanDerWijden** [comment]
I made it compliant with the spec, but the code is very ugly now. I really dislike that we have to pass the chainconfig to getBody and check if the block is post-shanghai

File: eth/catalyst/api.go [hunk 9]
```diff
@@ -756,3 +758,65 @@ func (api *ConsensusAPI) heartbeat() {
...
+// of block bodies by the engine api.
+func (api *ConsensusAPI) GetPayloadBodiesByRangeV1(start, count uint64) ([]*beacon.ExecutionPayloadBodyV1, error) {
+	if start == 0 || count == 0 || count > 1024 {
+		return nil, beacon.InvalidParams.With(fmt.Errorf("invalid start or count, start: %v count: %v", start, count))
+	}
+	current := api.eth.BlockChain().CurrentBlock().NumberU64()
+	// Return [] if the requested range is past our latest block
+	if current < start {
+		return []*beacon.ExecutionPayloadBodyV1{}, nil
+	}
```
Comment (holiman) [inline eth/catalyst/api.go:783]: The way I suggested it  you don't need this check

File: eth/catalyst/api.go [hunk 10]
```diff
@@ -756,3 +758,65 @@ func (api *ConsensusAPI) heartbeat() {
...
+
+	for j, tx := range body.Transactions {
+		data, _ := tx.MarshalBinary()
+		txs[j] = hexutil.Bytes(data)
+	}
+
+	// Post-shanghai withdrawals MUST be set to empty slice instead of nil
+	if withdrawals == nil && api.eth.BlockChain().Config().IsShanghai(block.Time()) {
+		withdrawals = make([]*types.Withdrawal, 0)
+	}
```
Comment (holiman) [inline eth/catalyst/api.go:816]: An alternative to checking against the config would be to check if the header has a `nil` withdrawalsHash or not

File: eth/catalyst/api.go (hunk 9 above)
Comment (holiman) [reply eth/catalyst/api.go:783]: In fact, you don't need it here either. 
```
current = 10
start = 20
count =4

end = 24
end = 20 // end = current
for i = 20; i < 20 ; i++ 
```

File: eth/catalyst/api.go [hunk 11]
```diff
@@ -756,3 +758,61 @@ func (api *ConsensusAPI) heartbeat() {
...
+		return nil, beacon.InvalidParams.With(fmt.Errorf("invalid start or count, start: %v count: %v", start, count))
+	}
+	// limit count up until current
+	current := api.eth.BlockChain().CurrentBlock().NumberU64()
+	end := start + count
+	if end > current {
+		end = current
+	}
+	var bodies []*beacon.ExecutionPayloadBodyV1
+	for i := uint64(start); i < end; i++ {
```
Comment (holiman) [inline eth/catalyst/api.go:786]:
```suggestion
	for i := start; i < end; i++ {
```
why cast uint64 to uint64?

**Review by holiman — APPROVED** [review]
Yeah lgtm now

**Pull request merged**